### PR TITLE
Make front:SourceFile include ScannedLine[]

### LINKED ISSUE #408: Make `front:SourceFile` include `ScannedLine[]`
This will enable:

- Provide columns in units of UTF-8 and UTF-16 code unit in addition to code points
- Support display of line in which error occurred with location highlighted (similar to clang)

(These things are not part of this issue.)

### PR DESCRIPTION
**Pull request by manuranga**
Make front:SourceFile include ScannedLine[]
Fixes #408

### CODE REVIEW

File: compiler/modules/front.syntax/scan.bal [hunk 1]
```diff
@@ -189,6 +189,13 @@ function unicodeEscapeValue(string fragment) returns string|error {
     return ch;
 }
 
+function scanLines(string[] lines) returns readonly & ScannedLine[] {
+    //JBUG NPE from BIROptimizer.java:159 when inlined
+    ScannedLine[] result = from var l in lines select scanLine(l);
+    // xxx don't clone, still haven't figured out how to use query to create readonly lists
+    return result.cloneReadOnly();
```
Comment (manuranga) [inline compiler/modules/front.syntax/scan.bal:196]: @jclark what would be a good way to do this ? Is cloning ok?

Comment (jclark) [reply compiler/modules/front.syntax/scan.bal:196]: cloneReadOnly is fine.

The way it ought to work is that a query expression creates a read-only list when the contextually expected type is readonly.

Comment (manuranga) [reply compiler/modules/front.syntax/scan.bal:196]: @gimantha please note.

**Comment by manuranga** [comment]
@jclark 
I can refactor to replace `LoadedSourcePart` with `SourceFile`. Is this a good refactoring?
If not, please consider this PR ready to be reviewed.

File: compiler/modules/front.syntax/parse.bal [hunk 2]
```diff
@@ -8,8 +8,8 @@ public type FilePath record {|
 |};
 
 public function parseModulePart(string[] lines, FilePath path, int partIndex) returns ModulePart|err:Syntax {
-    SourceFile file = new(path);
-    Tokenizer tok = new (lines, file);
+    SourceFile file = new(path, scanLines(lines));
```
Comment (jclark) [inline compiler/modules/front.syntax/parse.bal]: This should be factored into a function

function createSourceFile(string[] lines, FilePath path) returns SourceFile

File: compiler/modules/front.syntax/token.bal [hunk 3]
```diff
@@ -285,10 +285,12 @@ public readonly class SourceFile {
     *err:File;
     private string fn;
     private string? dir;
+    private ScannedLine[] lines;
 
-    public function init(FilePath path) {
+    public function init(FilePath path, readonly & ScannedLine[] lines) {
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Don't want this public any more.

**Comment by jclark** [comment]
> @jclark 
> 
> I can refactor to replace `LoadedSourcePart` with `SourceFile`. Is this a good refactoring?
> 
> If not, please consider this PR ready to be reviewed.

We may well want to do that. Not sure right now. It will become clear when we get a bit further along.

File: compiler/modules/front.syntax/tests/testParser.bal [hunk 4]
```diff
@@ -49,8 +49,8 @@ type SingleStringTokenizerTestCase [string, string];
     dataProvider: sourceFragments
 }
 function testTokenizer(string k, string[] lines) returns error? {
-    SourceFile file = new({ filename: k });
-    Tokenizer tok = new (lines, file);
+    SourceFile file = new({ filename: k }, scanLines(lines));
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/testParser.bal]: Can't we use createSourceFile here?

File: compiler/modules/front.syntax/tests/testParser.bal [hunk 5]
```diff
@@ -119,8 +119,8 @@ function reduceToWords(string k, string rule, string[] fragment) returns err:Syn
         modulePartToWords(w, check parseModulePart(fragment, { filename: k }, 0));
     }
     else {
-        SourceFile file = new({ filename: k });
-        Tokenizer tok = new (fragment, file);
+        SourceFile file = new({ filename: k }, scanLines(fragment));
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/testParser.bal]: And here?

File: compiler/modules/front.syntax/tests/tokenTests.bal [hunk 6]
```diff
@@ -47,8 +47,8 @@ function test2() {
 }
 
 function tokenization(string str) returns Token[]|error {
-    SourceFile file = new({ filename: "<internal>" });
-    Tokenizer tok = new([str], file);
+    SourceFile file = new({ filename: "<internal>" }, [scanLine(str)]);
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/tokenTests.bal]: And here?

File: compiler/modules/front.syntax/tests/testParser.bal (hunk 4 above)
Comment (manuranga) [reply compiler/modules/front.syntax/tests/testParser.bal]: Ah, I mistakenly though it can't be used here. Fixed.

**Pull request merged**